cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -660,36 +736,46 @@ struct ISOCalendar {
 
 template <
     template <typename...> class Op,
-    template <template <typename...> class OpExec, typename Duration, typename OutType>
+    template <template <typename...> class OpExec, typename Duration, typename InType, typename OutType>
     class ExecTemplate,
     typename OutType>
-std::shared_ptr<ScalarFunction> MakeTemporal(
+std::shared_ptr<ScalarFunction> MakeTemporal(bool enable_date,
```

Comment:
Instead of a bool, I think it'd be cleaner to split the date kernel registration into a separate function:

```cpp
template <
    template <typename...> class Op,
    template <template <typename...> class OpExec, typename Duration, typename InType, typename OutType>
    class ExecTemplate,
    typename OutType>
std::shared_ptr<ScalarFunction> AddDateKernels(ScalarFunction* func, const std::shared_ptr<arrow::DataType>& out_type, KernelInit init = nullptr) {
  {
    auto exec = ExecTemplate<Op, days, Date32Type, OutType>::Exec;
    DCHECK_OK(func->AddKernel({date32()}, out_Type, std::move(exec), init);
  }
  {
    auto exec = ExecTemplate<Op, std::chrono::milliseconds, Date64Type, OutType>::Exec;
    DCHECK_OK(func->AddKernel({date64()}, out_Type, std::move(exec), init);
  }
}
```

Then you can just call (for instance) `AddDateKernels<Year, TemporalComponentExtract, Int64Type>(year.get(), int64());` below.

Reply:
Thanks, but I suggest to keep the provided implementation in this PR:
- It requires less code/It change less parts.
- It integrates well into the `MakeTemporal*` functions.
- It is setting up the date kernels (optionally) at the moment of creating the function.
Also, I think that if I add `AddDateKernels` we would need 2 versions: one for MakeTemporal and other for MakeSimpleUnaryTemporal.
So I think it makes sense to keep the current code. Let me know what do you think here @lidavidm

Reply:
I will note though, a struct like this can collapse MakeTemporal and MakeSimpleUnaryTemporal into one function:

```cpp
template <template <typename...> class Op, typename Duration, typename InType>
struct SimpleUnaryTemporal {
  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
    return SimpleUnary<Op<Duration, InType>>(ctx, batch, out);
  }
};
```

You would then replace OutType with `typename... Args` in `MakeTemporal` and you could replace `MakeSimpleUnaryTemporal<Foo>` with `MakeTemporal<Foo, SimpleUnaryTemporal>`. Then you wouldn't need two versions of `AddDateKernels`.

Also note that the function creation is only done once on initialization, so it doesn't really matter whether it's done in the same function or not.

Reply:
Thanks, I added `/*enable_date=*/` to indicate the intention of the boolean parameter.
Regarding the `SimpleUnaryTemporal` struct, sounds like a nice idea for a future refactor, however if you feel strongly about carrying out these sort of changes in this PR, let me know please!